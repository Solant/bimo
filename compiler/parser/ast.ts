export enum AstNodeType {
    Structure = 'structure',
    Field = 'field',
    Document = 'document',
    SimpleType = 'simpletype',
    ParametrizedType = 'parametrizedtype',
    Number = 'number',
    Endianness = 'endianness',
    FieldRef = 'fieldref'
}

export interface NodePosition {
    line: number,
    column: number,
    offset: number,
}

export interface BaseAstNode {
    pos: NodePosition,
}

export interface StructureAstNode extends BaseAstNode {
    type: AstNodeType.Structure,
    default: boolean,
    name: string,
    fields: FieldAstNode[],
}

export interface FieldAstNode extends BaseAstNode {
    type: AstNodeType.Field,
    name: string,
    fieldType: SimpleFieldTypeAstNode | ParamFieldTypeAstNode,
}

export interface SimpleFieldTypeAstNode extends BaseAstNode {
    type: AstNodeType.SimpleType,
    typeName: string,
}

export interface ParamFieldTypeAstNode extends BaseAstNode {
    type: AstNodeType.ParametrizedType,
    typeName: string,
    typeArgs: Array<ParamFieldTypeAstNode | EndiannessLiteralAstNode | NumberLiteralAstNode | FieldRefAstNode>,
}

export interface DocumentAstNode extends BaseAstNode {
    type: AstNodeType.Document,
    structures: StructureAstNode[],
}

export interface NumberLiteralAstNode extends BaseAstNode {
    type: AstNodeType.Number,
    value: number,
}

export interface EndiannessLiteralAstNode extends BaseAstNode {
    type: AstNodeType.Endianness,
    value: string,
}

export interface FieldRefAstNode extends BaseAstNode {
    type: AstNodeType.FieldRef,
    fieldName: string,
}

export type AstNode = StructureAstNode
    | FieldAstNode
    | DocumentAstNode
    | SimpleFieldTypeAstNode
    | ParamFieldTypeAstNode
    | NumberLiteralAstNode
    | EndiannessLiteralAstNode
    | FieldRefAstNode;

export type BiMoAst = DocumentAstNode;
