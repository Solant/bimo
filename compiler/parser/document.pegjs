Document = _ structs:AnyStructure+ {
	return {
	    type: 'document',
	    pos: location().start,
	    structures: structs
    };
}

AnyStructure = DefaultStructureDeclaration / StructureDeclaration

StructureDeclaration = 'struct' _ name:TypeName _ '{' dec:Declaration+ _ '}' _ {
    return {
        type: 'structure',
        pos: location().start,
        name: name.join(''),
        fields: dec,
        default: false
    }
}

DefaultStructureDeclaration = 'default' _ struct:StructureDeclaration {
    return Object.assign({}, struct, { default: true });
}

TypeName = [a-zA-Z0-9]+

VarName = [a-zA-Z0-9]+

Type = ParametrizedType / SimpleType

SimpleType = typeName:TypeName {
	return {
	    type: 'simpletype',
	    pos: location().start,
	    typeName: typeName.join('')
    };
}

NumberLiteral = [0-9]+ {
	return {
	    type: 'number',
	    pos: location().start,
	    value: parseInt(text())
    };
}

EndiannessLiteral = val:('le' / 'be') {
	return {
    	type: 'endianness',
    	pos: location().start,
		value: text(),
    }
}

FieldRef = '#' fieldName:[a-zA-Z0-9]+ {
	return {
	    type: 'fieldref',
	    pos: location().start,
	    fieldName: fieldName.join(''),
    };
}

PossibleTypeArgs = FieldRef / ParametrizedType / NumberLiteral / EndiannessLiteral / SimpleType

TypeArg = _ arg:PossibleTypeArgs ','? _ {
	return arg;
}

ParametrizedType = typeName:TypeName '<' args:TypeArg+ '>' {
	return {
	    type: 'parametrizedtype',
	    pos: location().start,
	    typeName: typeName.join(''),
	    typeArgs: args
    }
}

Declaration = ComputedFieldDeclaration / FieldDeclaration

FieldDeclaration = _ variable:VarName _ ':' _ fieldType:Type _ ';' {
    return {
        type: 'field',
        pos: location().start,
        name: variable.join(''),
        fieldType: fieldType
    }
}

ComputedFieldDeclaration = _ variable:VarName _ ':' _ fieldType:Type _ '=' _ expr:ExpressionBody _ ';' {
    return {
        type: 'computedfield',
        pos: location().start,
        name: variable.join(''),
        fieldType: fieldType,
        expr: expr,
    }
}

ExpressionBody = additive

additive
  = first:multiplicative rest:(("+" / "-") multiplicative)+ {
    return rest.reduce(function(memo: any, curr: any) {
      return {
        tag: 'BinaryOperator',
        op: curr[0], left: memo, right: curr[1]
      };
    }, first);
  }
  / multiplicative

multiplicative
  = first:primary rest:(("*" / "/") primary)+ {
    return rest.reduce(function(memo: any, curr: any) {
      return {
        tag: 'BinaryOperator',
        op: curr[0], left: memo, right: curr[1]
      };
    }, first);
  }
  / primary

primary
  = number / fun / var
  / "(" additive:additive ")" { return additive; }

fun = name:[a-zA-Z]+ '(' body:var ')' {
 return {
     type: 'Function',
        name: name.join(''),
        body: body,
    }
}

var = [a-zA-Z0-9]+ {
    return { type: 'Var', value: text() }
}

number = digits:[0-9]+ {
    return { type: 'Number', value: parseInt(digits.join(""), 10) };
}

_ "whitespace" = [ \t\n\r]*

