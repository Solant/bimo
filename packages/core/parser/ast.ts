export enum AstNodeType {
    ComputedField = 'computedfield',
    Structure = 'structure',
    Field = 'field',
    Document = 'document',
    SimpleType = 'simpletype',
    ParametrizedType = 'parametrizedtype',
    Endianness = 'endianness',
    FieldRef = 'fieldref',
    Expression = 'Expression',
    BinaryOperator = 'BinaryOperator',
    Variable = 'Var',
    Number = 'Number',
    Function = 'Function',
}

export interface NodePosition {
    line: number,
    column: number,
    offset: number,
}

export interface BaseAstNode {
    pos: NodePosition,
}

export interface StructureAstNode extends BaseAstNode {
    type: AstNodeType.Structure,
    default: boolean,
    name: string,
    fields: Array<FieldAstNode | ComputedFieldAstNode>,
}

export interface FieldAstNode extends BaseAstNode {
    type: AstNodeType.Field,
    name: string,
    fieldType: SimpleFieldTypeAstNode | ParamFieldTypeAstNode,
}

export interface ComputedFieldAstNode extends BaseAstNode {
    type: AstNodeType.ComputedField,
    name: string,
    fieldType: SimpleFieldTypeAstNode | ParamFieldTypeAstNode,
    expr: Expression.Expression,
}

export interface SimpleFieldTypeAstNode extends BaseAstNode {
    type: AstNodeType.SimpleType,
    typeName: string,
}

export interface ParamFieldTypeAstNode extends BaseAstNode {
    type: AstNodeType.ParametrizedType,
    typeName: string,
    typeArgs: Array<ParamFieldTypeAstNode | EndiannessLiteralAstNode | NumberLiteralAstNode | FieldRefAstNode>,
    args: Expression.Expression[],
}

export interface DocumentAstNode extends BaseAstNode {
    type: AstNodeType.Document,
    structures: StructureAstNode[],
}

export interface NumberLiteralAstNode extends BaseAstNode {
    type: AstNodeType.Number,
    value: number,
}

export interface EndiannessLiteralAstNode extends BaseAstNode {
    type: AstNodeType.Endianness,
    value: string,
}

export interface FieldRefAstNode extends BaseAstNode {
    type: AstNodeType.FieldRef,
    fieldName: string,
}

export type AstNode = StructureAstNode
    | FieldAstNode
    | ComputedFieldAstNode
    | DocumentAstNode
    | SimpleFieldTypeAstNode
    | ParamFieldTypeAstNode
    | NumberLiteralAstNode
    | EndiannessLiteralAstNode
    | FieldRefAstNode
    | Expression.Expression
    | Expression.ExpressionNode;

export namespace Expression {
    export interface Variable {
        type: AstNodeType.Variable,
        value: string
    }

    export interface Number {
        type: AstNodeType.Number,
        value: number,
    }

    export interface Function {
        type: AstNodeType.Function,
        name: string,
        body: Variable,
    }

    export interface BinaryOperator {
        type: AstNodeType.BinaryOperator,
        op: '+' | '-' | '*' | '/',
        left: ExpressionNode,
        right: ExpressionNode,
    }

    export type ExpressionNode = Variable
        | Number
        | Function
        | BinaryOperator;

    export interface Expression extends BaseAstNode {
        type: AstNodeType.Expression,
        body: ExpressionNode
    }
}

export type BiMoAst = DocumentAstNode;
