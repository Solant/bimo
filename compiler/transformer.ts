import { BaseType, CustomType, TypeTag } from './types';
import {AstNode, AstNodeType, BiMoAst, NodePosition} from './parser/ast';

type DiscriminateUnion<T, K extends keyof T, V extends T[K]> = T extends Record<K, V> ? T : never;

interface EnterExitVisitor<T> {
    enter?: (node: T) => any,
    exit?: (node: T) => any,
}
type AstNodeVisitor = { [key in AstNodeType]?: EnterExitVisitor<DiscriminateUnion<AstNode, 'type', key>> };

function enterNode<T extends AstNode>(node: T, visitors: AstNodeVisitor[]) {
    for (let visitor of visitors) {
        // TODO: fix type casting
        // @ts-ignore
        const callback: EnterExitVisitor<T> = visitor[node.type];
        if (callback && callback.enter) {
            callback.enter(node);
        }
    }
}

function exitNode(node: AstNode, visitors: AstNodeVisitor[]) {
    for (let visitor of visitors) {
        // TODO: fix type casting
        // @ts-ignore
        const callback: EnterExitVisitor<T> = visitor[node.type];
        if (callback && callback.exit) {
            callback.exit(node);
        }
    }
}

function traverse(nodes: AstNode[], visitors: AstNodeVisitor[]) {
    nodes.forEach((node) => {
        switch (node.type) {
            case AstNodeType.Structure: {
                enterNode(node, visitors);
                traverse(node.fields, visitors);
                exitNode(node, visitors);
                break;
            }
            case AstNodeType.Field: {
                enterNode(node, visitors);
                exitNode(node, visitors);
                break;
            }
            case AstNodeType.Document: {
                enterNode(node, visitors);
                traverse(node.structures, visitors);
                exitNode(node, visitors);
                break;
            }
        }
    });
}

class CompileError extends Error {
    constructor(m: string, pos?: { line: number, column: number }) {
        if (pos) {
            super(`Compilation error at input:${pos.line}:${pos.column}\n${m}`)
        } else {
            super(m);
        }
    }
}

export function transform(ast: BiMoAst): BaseType[] {
    const output: BaseType[] = [];
    const visitors: AstNodeVisitor[] = [];

    // populate builtin types
    output.push({ tag: TypeTag.BuiltIn, name: 'i32le' });

    // unique default structure
    let defaultCounter = 0;
    let defaultPos: NodePosition | undefined = undefined;
    visitors.push({
        structure: {
            enter(node) {
                if (node.default) {
                    if (defaultCounter) {
                        throw new CompileError(`Default structure already defined at ${defaultPos}`, node.pos);
                    }
                    defaultCounter++;
                    defaultPos = node.pos;
                }
            }
        },
        document: {
            exit() {
                if (defaultCounter === 0) {
                    throw new CompileError(`No default structure found in the document`);
                }
            }
        }
    });

    // register structures
    let currentStruct: CustomType | undefined  = undefined;
    visitors.push({
        structure: {
            enter(node) {
                if (output.findIndex(t => t.name === node.name) !== -1) {
                    throw new CompileError(`Type ${node.name} was already declared`, node.pos);
                }
                currentStruct = {
                    tag: TypeTag.Custom,
                    default: node.default,
                    name: node.name,
                    props: [],
                };
                output.push(currentStruct);
            }
        },
        field: {
            enter(node) {
                const fields = currentStruct!.props;
                if (fields.findIndex(f => f.name === node.name) !== -1) {
                    throw new CompileError(`Field ${node.name} was already declared`, node.pos);
                }

                const fieldType = output.find(t => t.name === node.fieldType);
                if (!fieldType) {
                    throw new CompileError(`Type ${node.fieldType} is not declared`, node.pos);
                }
                fields.push({
                    name: node.name,
                    type: fieldType,
                });
            },
        },
    });

    traverse([ast], visitors);

    return output;
}
