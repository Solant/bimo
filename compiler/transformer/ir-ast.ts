import { TypeName } from '../builtInTypes';
import { Expression } from '../parser/ast';
import ExpressionNode = Expression.ExpressionNode;

export enum TypeTag {
    BuiltIn,
    Custom,
}

export interface BuiltInType {
    tag: TypeTag.BuiltIn,
    name: TypeName,
    typeArgs: TypeArgument,
}

export function isBuiltInType(t: BaseType): t is BuiltInType {
    return t.tag === TypeTag.BuiltIn;
}

export function isBuiltInArray(t: BuiltInType): boolean {
    return t.name === 'array';
}

type TypeField = Field | ComputedField;

export interface CustomType {
    tag: TypeTag.Custom,
    default: boolean,
    name: string,
    props: TypeField[],
}

export function isCustomType(t: BaseType): t is CustomType {
    return t.tag === TypeTag.Custom;
}

export type BaseType = CustomType | BuiltInType;

export interface TypeArgument {
    isLe?: boolean,
    lengthOf?: string,
    length?: number,
    type?: BaseType,
}

export interface Field {
    name: string,
    access: 'public' | 'private',
    type: BaseType,
    computed: false,
}

export interface ComputedField {
    name: string,
    access: 'public' | 'private',
    type: BaseType,
    computed: true,
    expression: Expression.ExpressionNode,
}