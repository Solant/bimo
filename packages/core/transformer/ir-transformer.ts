import { BaseType, ComputedField, CustomType, Field, isBuiltInArray, isBuiltInType, TypeTag } from './ir-ast';
import {
    AstNode,
    AstNodeType,
    BiMoAst,
    NodePosition,
    ParamFieldTypeAstNode,
    SimpleFieldTypeAstNode
} from '../parser/ast';
import { isTypeName } from '../builtInTypes';
import { assertNever, CompileError } from '../assertions';
import { enterNode, exitNode, GenericVisitor } from '../visitor';

type AstNodeVisitor = GenericVisitor<AstNodeType, AstNode, undefined, any>;


function pathFinder<T extends B, B extends { type: any }>(path: B[], type: any): T | undefined {
    const item = path.find(item => item.type === type);
    if (item) {
        return item as T;
    } else {
        return undefined;
    }
}

function traverse<T>(nodes: AstNode[], visitors: AstNodeVisitor[], path: AstNode[], scope: T) {
    nodes.forEach((node) => {
        const currentPath = [...path, node];
        const scope = undefined;
        if (node === null) {
            console.log('ooops');
        }
        switch (node.type) {
            case AstNodeType.Structure: {
                enterNode(node, visitors, currentPath, scope);
                traverse(node.fields, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.ComputedField: {
                enterNode(node, visitors, currentPath, scope);
                traverse([node.fieldType], visitors, currentPath, scope);
                traverse([node.expr], visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.BinaryOperator: {
                enterNode(node, visitors, currentPath, scope);
                traverse([node.left], visitors, currentPath, scope);
                traverse([node.right], visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.Function: {
                enterNode(node, visitors, currentPath, scope);
                traverse([node.body], visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.Variable: {
                enterNode(node, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.Field: {
                enterNode(node, visitors, currentPath, scope);
                traverse([node.fieldType], visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.Document: {
                enterNode(node, visitors, currentPath, scope);
                traverse(node.structures, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.ParametrizedType: {
                enterNode(node, visitors, currentPath, scope);
                traverse(node.typeArgs, visitors, currentPath, scope);
                // FIXME: nested type args
                traverse(node.args.filter(Boolean), visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.SimpleType: {
                enterNode(node, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.FieldRef: {
                enterNode(node, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.Number: {
                enterNode(node, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.Endianness: {
                enterNode(node, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.Expression: {
                enterNode(node, visitors, currentPath, scope);
                traverse([node.body], visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            case AstNodeType.String: {
                enterNode(node, visitors, currentPath, scope);
                exitNode(node, visitors, currentPath, scope);
                break;
            }
            default: {
                assertNever(node);
            }
        }
    });
}

export function transform(ast: BiMoAst): BaseType[] {
    const output: BaseType[] = [];
    const visitors: AstNodeVisitor[] = [];

    // unique default structure
    (function () {
        let defaultCounter = 0;
        let defaultPos: NodePosition | undefined = undefined;
        visitors.push({
            structure: {
                enter(node) {
                    if (node.default) {
                        if (defaultCounter) {
                            throw new CompileError(`Default structure already defined at ${defaultPos}`, node.pos);
                        }
                        defaultCounter++;
                        defaultPos = node.pos;
                    }
                }
            },
            document: {
                exit() {
                    if (defaultCounter === 0) {
                        throw new CompileError(`No default structure found in the document`);
                    }
                }
            }
        });
    })();

    class Stack<T> {
        data: T[];
        constructor() {
            this.data = [];
        }
        push(a: T) {
            this.data.push(a);
        }
        pop(): T {
            return this.data.pop()!;
        }
        head(): T {
            return this.data[this.data.length - 1];
        }
    }

    // register structures
    const structs = new Stack<CustomType>();
    const fields = new Stack<Field | ComputedField>();
    const types = new Stack<BaseType>();

    function pickBaseType(node: SimpleFieldTypeAstNode | ParamFieldTypeAstNode): BaseType {
        const name = node.typeName;
        if (isTypeName(name)) {
            return {
                tag: TypeTag.BuiltIn,
                name: name,
                typeArgs: {},
                args: [],
            };
        } else if (output.find(t => t.name === name)) {
            return output.find(t => t.name === name)!;
        } else {
            throw new CompileError(`Type ${name} is not declared`, node.pos);
        }
    }

    visitors.push({
        structure: {
            enter(node) {
                if (output.findIndex(t => t.name === node.name) !== -1) {
                    throw new CompileError(`Type ${node.name} was already declared`, node.pos);
                }
                structs.push({
                    tag: TypeTag.Custom,
                    default: node.default,
                    name: node.name,
                    props: [],
                });
                output.push(structs.head());
            },
            exit() {
                structs.pop();
            }
        },
        field: {
            enter(node) {
                const existingFields = structs.head().props;
                if (existingFields.findIndex(f => f.name === node.name) !== -1) {
                    throw new CompileError(`Field ${node.name} was already declared`, node.pos);
                }
                fields.push({
                    name: node.name,
                    access: 'public',
                    // @ts-ignore
                    type: {},
                });
                existingFields.push(fields.head());
            },
            exit() {
                fields.pop();
            }
        },
        [AstNodeType.ComputedField]: {
            enter(node) {
                const existingFields = structs.head().props;
                if (existingFields.findIndex(f => f.name === node.name) !== -1) {
                    throw new CompileError(`Field ${node.name} was already declared`, node.pos);
                }
                fields.push({
                    name: node.name,
                    access: 'private',
                    // @ts-ignore
                    type: {},
                    computed: true,
                    expression: {},
                });
                existingFields.push(fields.head());
            },
            exit(node) {
                fields.pop();
            },
        },
        [AstNodeType.Expression]: {
            enter(node) {
                const f = fields.head() as ComputedField;
                // FIXME: remove f.body null check by moving it to child visitor
                if (f.expression && Object.keys(f.expression).length === 0) {
                    f.expression = node;
                }
            }
        },
        parametrizedtype: {
            enter(node, path) {
                const t = pickBaseType(node);
                switch (path[path.length - 2].type) {
                    case AstNodeType.Field:
                    case AstNodeType.ComputedField:
                        fields.head().type = t;
                        break;
                    case AstNodeType.ParametrizedType: {
                        const type = types.head();
                        if (isBuiltInType(type)) {
                            type.typeArgs.type = t;
                        } else {
                            throw new CompileError('Custom types can\'t have type params', node.pos);
                        }
                    }
                }
                types.push(t);

                // Find type body to parse
                if (node.args.length && node.args[0] === null) {
                    return;
                }

                // FIXME: nested type args
                traverse(node.args, [{
                    [AstNodeType.Expression]: {
                        enter(node) {
                            const type = types.head();
                            if (isBuiltInType(type) && type.args.length === 0) {
                                type.args.push(node);
                            }
                        }
                    },
                }], [...path, node], undefined);
            },
            exit(node) {
                const t = types.head();
                if (isBuiltInType(t) && isBuiltInArray(t)) {
                    const typeArgs = t.typeArgs;
                    if (typeArgs.type === undefined) {
                        throw new CompileError(`Arrays expect child type argument`, node.pos);
                    }
                    // TODO: remove size and lengthOf
                    if (!typeArgs.length && typeArgs.lengthOf === undefined && t.args.length === 0) {
                        throw new CompileError(`Arrays expect size type argument`, node.pos);
                    }
                }
                types.pop();
            },
        },
        endianness: {
            enter(node) {
                const type = types.head();
                if (isBuiltInType(type)) {
                    if (isBuiltInArray(type)) {
                        throw new CompileError(`Arrays don't have endianness param`, node.pos);
                    }
                    type.typeArgs.isLe = node.value === 'le';
                }
            }
        },
        Number: {
            enter(node, path) {
                const type = types.head();

                if(!pathFinder<ParamFieldTypeAstNode, AstNode>(path, AstNodeType.ParametrizedType)) {
                    if (type && isBuiltInType(type)) {
                        type.typeArgs.length = node.value;
                    }
                }
            }
        },
        fieldref: {
            enter(node) {
                const type = types.head();
                if (isBuiltInType(type)) {
                    const referredField = structs.head().props.find(p => p.name === node.fieldName);
                    if (referredField === undefined) {
                        throw new CompileError(`Array size field should be declared before array`, node.pos);
                    }
                    referredField.access = 'private';
                    type.typeArgs.lengthOf = node.fieldName;
                }
            }
        },
    });

    traverse([ast], visitors, [], undefined);

    return output;
}
