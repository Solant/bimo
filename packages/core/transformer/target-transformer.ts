import { BaseType, ComputedField, CustomType, Field, isBuiltInArray, isCustomType, TypeTag } from './ir-ast';
import * as TargetAst from './target-ast';
import {
    ExpressionTag,
    MainReadFunctionDeclaration,
    MainWriteFunctionDeclaration,
    ReadArrayType,
    ReadBuiltInType,
    ReadCustomType,
    WriteArrayType,
    WriteBuiltInType,
    WriteCustomType
} from './target-ast';
import { Expression } from '../parser/ast';

function getTypeField(f: Field | ComputedField): TargetAst.TypeFieldDeclaration {
    return {
        tag: ExpressionTag.TypeFieldDeclaration,
        name: f.name,
        type: f.type,
        public: f.access === 'public',
    };
}

function getTypeDeclaration(type: CustomType): TargetAst.TypeDeclaration {
    return {
        tag: ExpressionTag.TypeDeclaration,
        name: type.name,
        fields: type.props.map(getTypeField),
    }
}

function getWriteExpr(id: string, type: BaseType, expr?: Expression.ExpressionNode): WriteArrayType | WriteBuiltInType | WriteCustomType {
    switch (type.tag) {
        case TypeTag.BuiltIn:
            if (isBuiltInArray(type)) {
                return {
                    tag: ExpressionTag.WriteArrayType,
                    id,
                    typeArg: type.typeArgs,
                    expression: type.args[0],
                    write: getWriteExpr(`${id}[i]`, type.typeArgs.type!),
                };
            }
            return {
                tag: ExpressionTag.WriteBuiltInType,
                id,
                type,
                expression: expr,
            };
        case TypeTag.Custom:
            return {
                tag: ExpressionTag.WriteCustomType,
                id,
                type,
            }
    }
}

function getWriteFunctionDeclaration(type: CustomType): TargetAst.FunctionDeclaration {
    const props: Array<TargetAst.WriteCustomType | TargetAst.WriteBuiltInType | TargetAst.WriteArrayType> = type.props.map(p => {
        if (p.computed) {
            return getWriteExpr(p.name, p.type, p.expression);
        }
        return getWriteExpr(p.name, p.type);
    });

    return {
        tag: ExpressionTag.FunctionDeclaration,
        id: `write${type.name}`,
        type: 'void',
        signature: {
            tag: ExpressionTag.FunctionSignature,
            params: [
                { tag: ExpressionTag.FunctionParameter, id: 'struct', type },
                { tag: ExpressionTag.FunctionParameter, id: 'stream', type: 'BimoStream' },
            ],
        },
        body: [...props],
    };
}

function getReadExpr(id: string, type: BaseType): ReadCustomType | ReadBuiltInType | ReadArrayType {
    switch (type.tag) {
        case TypeTag.BuiltIn:
            if (isBuiltInArray(type)) {
                const expr = type.args[0];
                return {
                    tag: ExpressionTag.ReadArrayType,
                    id,
                    sizeExpr: expr,
                    read: getReadExpr('temp', type.typeArgs.type!),
                    type: type,
                }
            }

            return {
                tag: ExpressionTag.ReadBuiltInType,
                id,
                type: type,
            };
        case TypeTag.Custom:
            return {
                tag: ExpressionTag.ReadCustomType,
                id,
                type,
            }
    }
}

function getReadFunctionDeclaration(type: CustomType): TargetAst.FunctionDeclaration {
    const props: Array<ReadCustomType | ReadBuiltInType | ReadArrayType> = type.props.map(prop => getReadExpr(prop.name, prop.type));

    const createType: TargetAst.CreateType = {
        tag: ExpressionTag.CreateType,
        type: type,
        id: 'ret',
    };

    const ret: TargetAst.ReturnStatement = {
        tag: ExpressionTag.ReturnStatement,
        id: 'ret',
    };

    return {
        tag: ExpressionTag.FunctionDeclaration,
        id: `read${type.name}`,
        type: `${type.name}`,
        signature: {
            tag: ExpressionTag.FunctionSignature,
            params: [{ tag: ExpressionTag.FunctionParameter, id: 'stream', type: 'BimoStream' }],
        },
        body: [...props, createType, ret],
    };
}

export function transform(types: BaseType[]) {
    const customTypes = types.filter(isCustomType);

    const mainType = types.filter(isCustomType).find(t => t.default)!;

    const mainRead: MainReadFunctionDeclaration = {
        tag: ExpressionTag.MainReadFunctionDeclaration,
        type: mainType,
    };

    const mainWrite: MainWriteFunctionDeclaration = {
        tag: ExpressionTag.MainWriteFunctionDeclaration,
        type: mainType,
    };

    const functions = [
        ...customTypes.map(getReadFunctionDeclaration),
        ...customTypes.map(getWriteFunctionDeclaration),
        mainRead,
        mainWrite,
    ];

    const result: TargetAst.Program = {
        tag: ExpressionTag.Program,
        declarations: customTypes.map(getTypeDeclaration),
        functions,
    };

    return result;
}
