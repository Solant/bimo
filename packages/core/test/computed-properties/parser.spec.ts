import { parse } from '../../parser/document';
import * as Ast from '../../parser/ast';
import { transform as transformIR } from '../../transformer/ir-transformer';
import { transform as transformTarget } from '../../transformer/target-transformer';
import { ComputedField, CustomType } from '../../transformer/ir-ast';

describe('Parser', function () {
    it('should parse', () => {
        const result: Ast.DocumentAstNode = parse(`
        default struct Test {
            length: i32 = 3+lengthof(body);
        }
        `);

        expect((result.structures[0].fields[0] as Ast.ComputedFieldAstNode).expr).toMatchSnapshot();
    });

    describe('IR transformer', () => {
        const res = parse(`
        default struct Test {
            length: i32 = 3+lengthof(body);
        }
        `);
        const types = transformIR(res);

        it('should create prop', () => {
            expect((types[0] as CustomType).props[0].name).toBe('length');
        });

        it('should add body', () => {
            expect(((types[0] as CustomType).props[0] as ComputedField).expression).toMatchSnapshot();
        });
    });

    describe('target transformer', () => {
        const res = parse(`
        default struct Test {
            length: i32 = 3+lengthof(body);
        }
        `);

        it('should create target AST', () => {
            expect(() => transformTarget(transformIR(res))).not.toThrow();
        });
    });
});
