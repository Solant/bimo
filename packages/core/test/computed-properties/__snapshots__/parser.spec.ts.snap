// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser IR transformer should add body 1`] = `
Object {
  "body": Object {
    "left": Object {
      "pos": Object {
        "column": 27,
        "line": 3,
        "offset": 57,
      },
      "type": "Number",
      "value": 3,
    },
    "op": "+",
    "right": Object {
      "body": Object {
        "type": "Var",
        "value": "body",
      },
      "name": "lengthof",
      "type": "Function",
    },
    "type": "BinaryOperator",
  },
  "pos": Object {
    "column": 27,
    "line": 3,
    "offset": 57,
  },
  "type": "Expression",
}
`;

exports[`Parser should parse 1`] = `
Object {
  "body": Object {
    "left": Object {
      "pos": Object {
        "column": 27,
        "line": 3,
        "offset": 57,
      },
      "type": "Number",
      "value": 3,
    },
    "op": "+",
    "right": Object {
      "body": Object {
        "type": "Var",
        "value": "body",
      },
      "name": "lengthof",
      "type": "Function",
    },
    "type": "BinaryOperator",
  },
  "pos": Object {
    "column": 27,
    "line": 3,
    "offset": 57,
  },
  "type": "Expression",
}
`;
